File: mesh/sdk/mesh.ts
/**
 * mesh.ts — UDS-Mesh TypeScript SDK for emitting Λ-signed cross-organ OTEL spans.
 *
 * Functional mirror of mesh.py. Layer 5 (Observability) of the 7-layer SZL
 * architecture. Lets any organ (a11oy, sentra, amaru, killinchu, rosie) emit spans
 * conforming to ../../schemas/spans/*.yaml carrying the szl.mesh.* governance envelope.
 *
 * REAL behaviour (no mocks):
 *   - W3C Trace Context (https://www.w3.org/TR/trace-context/) parse/format.
 *   - DSSE PAE v1 receipt (HMAC-SHA-256 dev signer; same PAE as formula_receipts.py).
 *   - BLS-style aggregate verification modelling lutar-lean #180 (sorry-free)
 *     `Lutar.Round11.BLS.aggregate_verify`: Σ σ_i = (Σ sk_i)·h, so a batch of N
 *     mesh receipts verifies in ONE aggregate check. Production swaps the additive
 *     prime-field model for blst BLS12-381 — the verify contract is identical.
 *     Runtime coordinate: szl-holdings/amaru/szl_bls_aggregate.py.
 *
 * BigInt is used for the 256-bit scalar field; no external dependency.
 *
 * SPDX-License-Identifier: Apache-2.0
 * Author: Yachay (CTO authority) · Built by Perplexity Computer Agent · SZL Holdings
 * Doctrine v11 LOCKED — 749 / 14 / 163.
 */
import { createHash, createHmac, randomBytes } from "node:crypto";

export const LAMBDA_FLOOR_MESH = 0.1;
export const ORGANS = ["a11oy", "sentra", "amaru", "killinchu", "rosie"] as const;
export type Organ = (typeof ORGANS)[number];

export const SPAN_NAMES: Record<Organ, readonly string[]> = {
  a11oy: ["a11oy.graph.lambda", "a11oy.graph.automorphism", "a11oy.graph.position"],
  sentra: ["sentra.gate.evaluate", "sentra.gate.attest", "sentra.gate.fail_closed"],
  amaru: ["amaru.sync.merge", "amaru.sync.receipt", "amaru.sync.drift_alert"],
  killinchu: ["killinchu.courier.dispatch", "killinchu.courier.deliver", "killinchu.courier.verify"],
  rosie: ["rosie.decision.evaluate", "rosie.decision.witness", "rosie.decision.replay"],
};

const HMAC_KEY = process.env.MESH_HMAC_KEY ?? "szl-mesh-hmac-dev-v1";

// ── W3C Trace Context ────────────────────────────────────────────────────────
const TRACEPARENT_RE =
  /^([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/;

export class TraceContext {
  constructor(
    public traceId: string,
    public spanId: string,
    public flags = "01",
    public version = "00",
  ) {}

  static newRoot(): TraceContext {
    return new TraceContext(randomBytes(16).toString("hex"), randomBytes(8).toString("hex"));
  }

  static parse(traceparent: string): TraceContext {
    const m = TRACEPARENT_RE.exec(traceparent.trim());
    if (!m) throw new Error(`invalid W3C traceparent: ${traceparent}`);
    if (m[2] === "0".repeat(32) || m[3] === "0".repeat(16))
      throw new Error("traceparent contains all-zero id (invalid per W3C)");
    return new TraceContext(m[2], m[3], m[4], m[1]);
  }

  traceparent(): string {
    return `${this.version}-${this.traceId}-${this.spanId}-${this.flags}`;
  }

  child(): TraceContext {
    return new TraceContext(this.traceId, randomBytes(8).toString("hex"), this.flags, this.version);
  }
}

// ── DSSE PAE v1 ──────────────────────────────────────────────────────────────
function pae(payloadType: string, payload: Buffer): Buffer {
  const t = Buffer.from(payloadType);
  return Buffer.concat([
    Buffer.from(`DSSEv1 ${t.length} `), t,
    Buffer.from(` ${payload.length} `), payload,
  ]);
}

function sign(payloadType: string, payload: Buffer, key = HMAC_KEY): string {
  return createHmac("sha256", key).update(pae(payloadType, payload)).digest("base64");
}

// ── BLS-style aggregate signer (models Lutar.Round11.BLS.aggregate_verify) ────
// secp256k1 group order as the additive scalar-field modulus.
const P = BigInt("0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141");

function scalar(material: Buffer | string): bigint {
  const buf = typeof material === "string" ? Buffer.from(material) : material;
  return BigInt("0x" + createHash("sha256").update(buf).digest("hex")) % P;
}

const BLS = {
  skOf(signerId: string): bigint {
    return scalar("sk:" + signerId);
  },
  sign(signerId: string, msgHash: bigint): bigint {
    return (this.skOf(signerId) * msgHash) % P;
  },
  aggregate(sigs: bigint[]): bigint {
    return sigs.reduce((a, s) => (a + s) % P, 0n);
  },
  // (Σ sk_i)·h == Σ σ_i  ← agg_sig_eq_agg_key_sig
  aggregateVerify(signerIds: string[], msgHash: bigint, aggSig: bigint): boolean {
    const aggKey = signerIds.reduce((a, sid) => (a + this.skOf(sid)) % P, 0n);
    return (aggKey * msgHash) % P === ((aggSig % P) + P) % P;
  },
};

// ── Merkle root (RFC 6962, duplicate-last padding) ───────────────────────────
function merkleRoot(leaves: string[]): string {
  if (leaves.length === 0) return "";
  let level = leaves.map((h) =>
    h.length === 64 ? Buffer.from(h, "hex") : createHash("sha256").update(h).digest(),
  );
  while (level.length > 1) {
    if (level.length % 2) level.push(level[level.length - 1]);
    const next: Buffer[] = [];
    for (let i = 0; i < level.length; i += 2)
      next.push(createHash("sha256").update(Buffer.concat([level[i], level[i + 1]])).digest());
    level = next;
  }
  return level[0].toString("hex");
}

// ── Span + emitter ────────────────────────────────────────────────────────────
export interface OtelMeshSpan {
  trace_id: string;
  span_id: string;
  parent_span_id?: string;
  name: string;
  start_time: string;
  end_time: string;
  status: { code: string };
  attributes: Record<string, string | number | boolean>;
}

export interface BatchAggregate {
  count: number;
  agg_sig: string; // hex (bigint)
  root: string;
  verified: boolean;
  span_ids: string[];
}

interface BatchEntry {
  otel: OtelMeshSpan;
  receiptHash: string;
}

export class MeshEmitter {
  private batch: BatchEntry[] = [];
  readonly signerId: string;

  constructor(public readonly organ: Organ, signerId?: string) {
    if (!ORGANS.includes(organ)) throw new Error(`unknown organ ${organ}`);
    this.signerId = signerId ?? `mesh:${organ}`;
  }

  emit(
    name: string,
    lambdaValue: number,
    opts: {
      trace?: TraceContext;
      attributes?: Record<string, string | number | boolean>;
      parentSpanId?: string;
      status?: string;
    } = {},
  ): OtelMeshSpan {
    if (!SPAN_NAMES[this.organ].includes(name))
      throw new Error(`span ${name} not in schema for organ ${this.organ}`);
    const trace = opts.trace ?? TraceContext.newRoot();
    const now = new Date().toISOString().replace("Z", "000Z");
    const drift = lambdaValue < LAMBDA_FLOOR_MESH || lambdaValue > 1.0;
    const receiptBody = JSON.stringify({
      attributes: opts.attributes ?? {},
      lambda: Math.round(lambdaValue * 1e6) / 1e6,
      name,
      organ: this.organ,
      span_id: trace.spanId,
      status: opts.status ?? "OK",
      trace_id: trace.traceId,
    });
    // NOTE: key order matches mesh.py json.dumps(sort_keys=True) for cross-SDK parity.
    const sortedBody = canonicalJSON({
      organ: this.organ,
      name,
      trace_id: trace.traceId,
      span_id: trace.spanId,
      lambda: Math.round(lambdaValue * 1e6) / 1e6,
      attributes: opts.attributes ?? {},
      status: opts.status ?? "OK",
    });
    const receiptHash = createHash("sha256").update(sortedBody).digest("hex");
    const signature = sign("application/vnd.in-toto+json", Buffer.from(sortedBody));
    void receiptBody;
    const otel: OtelMeshSpan = {
      trace_id: trace.traceId,
      span_id: trace.spanId,
      name,
      start_time: now,
      end_time: now,
      status: { code: opts.status ?? "OK" },
      attributes: {
        "szl.mesh.organ": this.organ,
        "szl.mesh.receipt_hash": receiptHash,
        "szl.mesh.dsse_payload_type": "application/vnd.in-toto+json",
        "szl.mesh.lambda_value": lambdaValue.toFixed(6),
        "szl.mesh.governance_drift": drift,
        "szl.mesh.dsse_signature": signature,
        ...(opts.attributes ?? {}),
      },
    };
    if (opts.parentSpanId) otel.parent_span_id = opts.parentSpanId;
    this.batch.push({ otel, receiptHash });
    return otel;
  }

  batchAggregate(): BatchAggregate {
    if (this.batch.length === 0)
      return { count: 0, agg_sig: "0", root: "", verified: true, span_ids: [] };
    const leaves = this.batch.map((b) => b.receiptHash);
    const root = merkleRoot(leaves);
    const rootHash = scalar(root);
    const sigs = this.batch.map(() => BLS.sign(this.signerId, rootHash));
    const agg = BLS.aggregate(sigs);
    const signers = this.batch.map(() => this.signerId);
    return {
      count: this.batch.length,
      agg_sig: agg.toString(16),
      root,
      verified: BLS.aggregateVerify(signers, rootHash, agg),
      span_ids: this.batch.map((b) => b.otel.span_id),
    };
  }

  drain(): OtelMeshSpan[] {
    const out = this.batch.map((b) => b.otel);
    this.batch = [];
    return out;
  }
}

export function verifyBatch(spans: OtelMeshSpan[], agg: BatchAggregate, signerId: string): boolean {
  const leaves = spans.map((s) => s.attributes["szl.mesh.receipt_hash"] as string);
  if (merkleRoot(leaves) !== agg.root) return false;
  const rootHash = scalar(agg.root);
  const signers = spans.map(() => signerId);
  return BLS.aggregateVerify(signers, rootHash, BigInt("0x" + agg.agg_sig));
}

// Canonical JSON with sorted keys, matching Python json.dumps(sort_keys=True, separators=(",",":")).
function canonicalJSON(obj: unknown): string {
  if (obj === null || typeof obj !== "object") return JSON.stringify(obj);
  if (Array.isArray(obj)) return "[" + obj.map(canonicalJSON).join(",") + "]";
  const keys = Object.keys(obj as Record<string, unknown>).sort();
  return "{" + keys.map((k) => JSON.stringify(k) + ":" + canonicalJSON((obj as Record<string, unknown>)[k])).join(",") + "}";
}

File: mesh/sdk/mesh.py
"""
mesh.py — UDS-Mesh Python SDK for emitting Λ-signed cross-organ OTEL spans.

Layer 5 (Observability) of the 7-layer SZL architecture. This SDK lets any of
the five organs (a11oy, sentra, amaru, killinchu, rosie) emit spans that conform
to the schemas in ../../schemas/spans/*.yaml and carry the cross-organ DSSE
governance-receipt envelope (`szl.mesh.*` attributes).

What is REAL here (no mocks):
  * W3C Trace Context propagation — `traceparent` parse/format per the W3C REC
    (https://www.w3.org/TR/trace-context/), version-00, 32-hex trace-id /
    16-hex span-id / 8-bit flags.
  * DSSE PAE v1 receipt over each span (HMAC-SHA-256 dev signer; same PAE scheme
    as formula_receipts.py and rosie-operator-console).
  * BLS-style aggregate verification over a batch of receipt signatures, modelling
    the round11 frontier formula `Lutar.Round11.BLS.aggregate_verify`
    (lutar-lean open PR #180, sorry-free): the aggregate of per-signer signatures
    equals the signature under the aggregate key, so a whole batch of mesh receipts
    verifies with one aggregate check instead of N. The additive-homomorphism model
    here is the runtime coordinate of `szl-holdings/amaru/szl_bls_aggregate.py`.

Honest crypto boundary: the production mesh uses BLS12-381 (py_ecc / blst). To keep
this SDK dependency-free and importable in CI, the aggregation is implemented over a
prime-field additive homomorphism that exercises the SAME algebraic identity proved
in Lean (Σ σ_i  vs  (Σ sk_i)·h). Swap `_BLSBackend` for blst in production; the
`aggregate_verify` contract is identical. This is disclosed, not hidden (HR-6).

SPDX-License-Identifier: Apache-2.0
Author: Yachay (CTO authority) · Built by Perplexity Computer Agent · SZL Holdings
Doctrine v11 LOCKED — 749 / 14 / 163.
"""
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import re
import secrets
import time
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from typing import Any, Iterable, Optional

# ── Doctrine constants ───────────────────────────────────────────────────────
LAMBDA_FLOOR_MESH = 0.10          # mesh governance floor (sentra.gate caveat)
ORGANS = ("a11oy", "sentra", "amaru", "killinchu", "rosie")
SPAN_NAMES = {
    "a11oy":     ("a11oy.graph.lambda", "a11oy.graph.automorphism", "a11oy.graph.position"),
    "sentra":    ("sentra.gate.evaluate", "sentra.gate.attest", "sentra.gate.fail_closed"),
    "amaru":     ("amaru.sync.merge", "amaru.sync.receipt", "amaru.sync.drift_alert"),
    "killinchu": ("killinchu.courier.dispatch", "killinchu.courier.deliver", "killinchu.courier.verify"),
    "rosie":     ("rosie.decision.evaluate", "rosie.decision.witness", "rosie.decision.replay"),
}

_HMAC_KEY = os.environ.get("MESH_HMAC_KEY", "szl-mesh-hmac-dev-v1").encode()

# ── W3C Trace Context (RFC: https://www.w3.org/TR/trace-context/) ─────────────
_TRACEPARENT_RE = re.compile(
    r"^(?P<version>[0-9a-f]{2})-(?P<trace_id>[0-9a-f]{32})-"
    r"(?P<span_id>[0-9a-f]{16})-(?P<flags>[0-9a-f]{2})$"
)


@dataclass
class TraceContext:
    trace_id: str
    span_id: str
    flags: str = "01"          # 01 = sampled
    version: str = "00"

    @classmethod
    def new_root(cls) -> "TraceContext":
        return cls(trace_id=secrets.token_hex(16), span_id=secrets.token_hex(8))

    @classmethod
    def parse(cls, traceparent: str) -> "TraceContext":
        m = _TRACEPARENT_RE.match(traceparent.strip())
        if not m:
            raise ValueError(f"invalid W3C traceparent: {traceparent!r}")
        if m["trace_id"] == "0" * 32 or m["span_id"] == "0" * 16:
            raise ValueError("traceparent contains all-zero id (invalid per W3C)")
        return cls(m["trace_id"], m["span_id"], m["flags"], m["version"])

    def traceparent(self) -> str:
        return f"{self.version}-{self.trace_id}-{self.span_id}-{self.flags}"

    def child(self) -> "TraceContext":
        """New child span in the same trace (fresh span-id)."""
        return TraceContext(self.trace_id, secrets.token_hex(8), self.flags, self.version)


# ── DSSE PAE v1 receipt (matches formula_receipts.py) ─────────────────────────
def _pae(payload_type: str, payload: bytes) -> bytes:
    t = payload_type.encode()
    return (b"DSSEv1 " + str(len(t)).encode() + b" " + t + b" "
            + str(len(payload)).encode() + b" " + payload)


def _sign(payload_type: str, payload: bytes, key: bytes = _HMAC_KEY) -> str:
    return base64.b64encode(hmac.new(key, _pae(payload_type, payload), hashlib.sha256).digest()).decode()


# ── BLS-style aggregate signer (models Lutar.Round11.BLS.aggregate_verify) ────
class _BLSBackend:
    """Additive-homomorphic model of BLS aggregation over a large prime field.

    Exercises the identity proved sorry-free in lutar-lean #180:
        aggSig(sk, h)      = Σ_i (sk_i · h)
        aggKeySig(sk, h)   = (Σ_i sk_i) · h
        agg_sig_eq_agg_key_sig : aggSig = aggKeySig
    so verifying a batch of N receipts is ONE aggregate check, not N.
    Production swaps this for blst/py_ecc BLS12-381; the contract is unchanged.
    """
    # A 256-bit prime (secp256k1 group order) used as the scalar field modulus.
    P = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141

    @staticmethod
    def _scalar(material: bytes) -> int:
        return int.from_bytes(hashlib.sha256(material).digest(), "big") % _BLSBackend.P

    @classmethod
    def sk_of(cls, signer_id: str) -> int:
        return cls._scalar(b"sk:" + signer_id.encode())

    @classmethod
    def sign(cls, signer_id: str, msg_hash: int) -> int:
        # σ_i = sk_i · h   (scalar mult in the additive model)
        return (cls.sk_of(signer_id) * msg_hash) % cls.P

    @classmethod
    def aggregate(cls, sigs: Iterable[int]) -> int:
        agg = 0
        for s in sigs:
            agg = (agg + s) % cls.P
        return agg

    @classmethod
    def aggregate_verify(cls, signer_ids: list[str], msg_hash: int, agg_sig: int) -> bool:
        # (Σ sk_i) · h  ==  Σ σ_i   ← agg_sig_eq_agg_key_sig
        agg_key = 0
        for sid in signer_ids:
            agg_key = (agg_key + cls.sk_of(sid)) % cls.P
        return (agg_key * msg_hash) % cls.P == agg_sig % cls.P


# ── Span + emitter ─────────────────────────────────────────────────────────────
@dataclass
class MeshSpan:
    organ: str
    name: str
    trace: TraceContext
    lambda_value: float
    attributes: dict[str, Any] = field(default_factory=dict)
    start_time: str = ""
    end_time: str = ""
    status: str = "OK"
    parent_span_id: Optional[str] = None
    receipt_hash: str = ""
    signature: str = ""

    def governance_drift(self) -> bool:
        return self.lambda_value < LAMBDA_FLOOR_MESH or self.lambda_value > 1.0

    def to_otel(self) -> dict:
        """Render to the OTEL + szl.mesh.* envelope defined by the span schemas."""
        d = {
            "trace_id": self.trace.trace_id,
            "span_id": self.trace.span_id,
            "name": self.name,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "status": {"code": self.status},
            "attributes": {
                "szl.mesh.organ": self.organ,
                "szl.mesh.receipt_hash": self.receipt_hash,
                "szl.mesh.dsse_payload_type": "application/vnd.in-toto+json",
                "szl.mesh.lambda_value": f"{self.lambda_value:.6f}",
                "szl.mesh.governance_drift": self.governance_drift(),
                **self.attributes,
            },
        }
        if self.parent_span_id:
            d["parent_span_id"] = self.parent_span_id
        return d


class MeshEmitter:
    """Emit Λ-signed spans for one organ and BLS-aggregate their receipts."""

    def __init__(self, organ: str, signer_id: Optional[str] = None):
        if organ not in ORGANS:
            raise ValueError(f"unknown organ {organ!r}; expected one of {ORGANS}")
        self.organ = organ
        self.signer_id = signer_id or f"mesh:{organ}"
        self._batch: list[tuple[MeshSpan, int, int]] = []  # (span, msg_hash, sig)

    def _receipt_payload(self, span: MeshSpan) -> bytes:
        body = {
            "organ": span.organ, "name": span.name,
            "trace_id": span.trace.trace_id, "span_id": span.trace.span_id,
            "lambda": round(span.lambda_value, 6),
            "attributes": span.attributes, "status": span.status,
        }
        return json.dumps(body, sort_keys=True, separators=(",", ":")).encode()

    def emit(self, name: str, lambda_value: float,
             trace: Optional[TraceContext] = None,
             attributes: Optional[dict] = None,
             parent_span_id: Optional[str] = None,
             status: str = "OK") -> MeshSpan:
        if name not in SPAN_NAMES[self.organ]:
            raise ValueError(f"span {name!r} not in schema for organ {self.organ}")
        trace = trace or TraceContext.new_root()
        now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")
        span = MeshSpan(
            organ=self.organ, name=name, trace=trace, lambda_value=lambda_value,
            attributes=attributes or {}, start_time=now, end_time=now,
            status=status, parent_span_id=parent_span_id,
        )
        payload = self._receipt_payload(span)
        span.receipt_hash = hashlib.sha256(payload).hexdigest()
        span.signature = _sign("application/vnd.in-toto+json", payload)
        msg_hash = _BLSBackend._scalar(span.receipt_hash.encode())
        sig = _BLSBackend.sign(self.signer_id, msg_hash)
        self._batch.append((span, msg_hash, sig))
        return span

    def batch_aggregate(self) -> dict:
        """Aggregate all receipts in the current batch into ONE BLS signature.

        Models lutar-lean #180 `aggregate_verify`: a verifier checks the whole
        batch with a single aggregate verification rather than N per-span checks.
        Spans in a batch share one common message hash (the Merkle root of the
        batch receipt hashes) so the algebraic identity applies directly.
        """
        if not self._batch:
            return {"count": 0, "agg_sig": 0, "verified": True, "root": ""}
        leaves = [s.receipt_hash for s, _, _ in self._batch]
        root = _merkle_root(leaves)
        root_hash = _BLSBackend._scalar(root.encode())
        # Re-sign each receipt under the shared batch root (one common h).
        signers = [self.signer_id] * len(self._batch)
        sigs = [_BLSBackend.sign(self.signer_id, root_hash) for _ in self._batch]
        agg = _BLSBackend.aggregate(sigs)
        verified = _BLSBackend.aggregate_verify(signers, root_hash, agg)
        return {"count": len(self._batch), "agg_sig": agg, "root": root,
                "verified": verified, "span_ids": [s.trace.span_id for s, _, _ in self._batch]}

    def drain(self) -> list[dict]:
        out = [s.to_otel() for s, _, _ in self._batch]
        self._batch.clear()
        return out


def _merkle_root(leaves: list[str]) -> str:
    """Binary SHA-256 Merkle root (RFC 6962 style, duplicate-last padding)."""
    if not leaves:
        return ""
    level = [bytes.fromhex(h) if len(h) == 64 else hashlib.sha256(h.encode()).digest()
             for h in leaves]
    while len(level) > 1:
        if len(level) % 2:
            level.append(level[-1])
        level = [hashlib.sha256(level[i] + level[i + 1]).digest()
                 for i in range(0, len(level), 2)]
    return level[0].hex()


def verify_batch(span_otel: list[dict], agg: dict, signer_id: str) -> bool:
    """Independent verifier: recompute the Merkle root from emitted spans and
    confirm the aggregate BLS signature covers exactly those receipt hashes."""
    leaves = [s["attributes"]["szl.mesh.receipt_hash"] for s in span_otel]
    if _merkle_root(leaves) != agg.get("root", ""):
        return False
    root_hash = _BLSBackend._scalar(agg["root"].encode())
    signers = [signer_id] * agg["count"]
    return _BLSBackend.aggregate_verify(signers, root_hash, agg["agg_sig"])


# ── Formula wiring (real-edge-v2): the mesh runtime INVOKES these formulas ─────
# Wired in so the bundle actually CALLS the formulas, not just ships their code:
#   * pac_bayes_quorum.gate_merge  — run BEFORE merging cross-organ verdicts
#   * bls_aggregate.cosign_chain   — REAL BLS12-381 aggregate over the Khipu chain root
#   * welford_streaming.StreamingStats — running stats on the OTLP trace fan-out
# Each carries a thesis-v22 citation + a real Lean theorem permalink (see the modules).
try:  # package-relative when imported as mesh.sdk.mesh
    from ..formulas import pac_bayes_quorum as _pbq
    from ..formulas import bls_aggregate as _blsagg
    from ..formulas import welford_streaming as _wstream
    _FORMULAS_OK = True
except Exception:  # pragma: no cover — flat import fallback
    try:
        import os as _os, sys as _sys
        _root = _os.path.dirname(_os.path.dirname(_os.path.dirname(_os.path.abspath(__file__))))
        if _root not in _sys.path:
            _sys.path.insert(0, _root)
        from mesh.formulas import pac_bayes_quorum as _pbq  # type: ignore
        from mesh.formulas import bls_aggregate as _blsagg  # type: ignore
        from mesh.formulas import welford_streaming as _wstream  # type: ignore
        _FORMULAS_OK = True
    except Exception as _fe:  # pragma: no cover
        _pbq = _blsagg = _wstream = None
        _FORMULAS_OK = False


class MeshGovernance:
    """Runtime coordinator that operationalises the three mesh formulas.

    The mesh creates ONE of these per cross-organ decision cycle. It:
      1. gates the merge with PAC-Bayes confidence + Byzantine quorum (gate_merge),
      2. folds each organ's fan-out latency through the Welford streaming accumulator,
      3. produces a REAL BLS12-381 aggregate signature over the Khipu chain root when
         3+ organs co-sign — a single FastAggregateVerify replaces N pairing checks.
    """

    def __init__(self, f: int = 1, confidence_floor: float = 0.5):
        if not _FORMULAS_OK:
            raise RuntimeError("mesh.formulas not importable; cannot run MeshGovernance")
        self.f = f
        self.confidence_floor = confidence_floor
        self.fanout = _wstream.StreamingStats()

    def observe_fanout(self, latency_ms: float) -> dict:
        """Fold one OTLP fan-out latency into the running Welford stats."""
        return self.fanout.observe(float(latency_ms))

    def gate(self, verdicts: dict, n_obs: int = 64, kl: float = 0.0,
             delta: float = 0.05) -> dict:
        """PAC-Bayes + Byzantine quorum gate, run BEFORE the mesh merges verdicts."""
        return _pbq.gate_merge(verdicts, f=self.f, n_obs=n_obs, kl=kl, delta=delta,
                               confidence_floor=self.confidence_floor)

    def cosign(self, organs: list[str], chain_root: str) -> dict:
        """REAL BLS12-381 aggregate signature over the cross-organ Khipu chain root."""
        return _blsagg.cosign_chain(organs, chain_root)

    def decide_and_cosign(self, verdicts: dict, chain_root: str,
                          n_obs: int = 64) -> dict:
        """End-to-end: gate the merge, then (if allowed) BLS-aggregate the co-signature."""
        gate = self.gate(verdicts, n_obs=n_obs)
        result = {"gate": gate, "fanout_stats": self.fanout.snapshot()}
        if gate["allow_merge"]:
            organs = gate["quorum"]["agreeing_organs"]
            result["cosign"] = self.cosign(organs, chain_root)
            result["merged_decision"] = gate["merged_decision"]
        else:
            result["cosign"] = None
            result["merged_decision"] = None
        result["value"] = bool(gate["allow_merge"])
        result["doctrine"] = "v11 · \u039b = Conjecture 1 (NEVER a theorem)"
        return result

    @staticmethod
    def formulas_index() -> list[dict]:
        """Honest index of the wired mesh formulas + Lean permalinks."""
        return [
            {"name": "pac_bayes_quorum", "citation": _pbq.CITATION,
             "lean_theorem": _pbq.LEAN_THEOREM,
             "lean_permalink": _pbq.LEAN_PERMALINK_PACBAYES},
            {"name": "bls_aggregate", "citation": _blsagg.CITATION,
             "lean_theorem": _blsagg.LEAN_THEOREM,
             "lean_permalink": _blsagg.LEAN_PERMALINK,
             "bls_available": _blsagg.AVAILABLE},
            {"name": "welford_streaming", "citation": _wstream.CITATION,
             "lean_theorem": _wstream.LEAN_THEOREM,
             "lean_permalink": _wstream.LEAN_PERMALINK},
        ]


__all__ = [
    "TraceContext", "MeshSpan", "MeshEmitter", "verify_batch",
    "MeshGovernance",
    "LAMBDA_FLOOR_MESH", "ORGANS", "SPAN_NAMES",
]
